host/sampi/http.go

```diff
@@ -0,0 +1,211 @@
+		return
+	}
+	if h.ID == "" {
+		rh.Error(errors.New("sampi: host id must not be blank"))
+		return
+	}
+
+	ch := make(chan *host.Job)
+
+	c.state.Begin()
+	if c.state.HostExists(h.ID) {
+		c.state.Rollback()
+		rh.Error(errors.New("sampi: host exists"))
+		return
+	}
+	c.state.AddHost(h, ch)
+	c.state.Commit()
+	go c.state.sendEvent(h.ID, "add")
+
+	// "defer" cleanups in a goroutine that waits until the http stream is closed.
+	go func() {
+		<-w.(http.CloseNotifier).CloseNotify()
+		c.state.Begin()
+		c.state.RemoveHost(h.ID)
+		c.state.Commit()
+		go c.state.sendEvent(h.ID, "remove")
+	}()
+
+	wr := sse.NewSSEWriter(w)
+	enc := json.NewEncoder(wr)
+	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
+	w.WriteHeader(200)
+	wr.Flush()
+	for data := range ch {
+		enc.Encode(data)
+		wr.Flush()
+	}
+}
+
+func addJobs(c *Cluster, w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	rh := httphelper.NewReponseHelper(w)
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		rh.Error(err)
+		return
+	}
+
+	var req map[string][]*host.Job
+	if err := json.Unmarshal(data, &req); err != nil {
+		rh.Error(err)
+		return
+	}
+	res, err := c.AddJobs(req)
+	if err != nil {
+		rh.Error(err)
+		return
+	}
+	rh.JSON(200, res)
+}
+
+func removeJob(c *Cluster, w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	hostID := ps.ByName("host_id")
+	jobIDs := []string{ps.ByName("job_id")}
+	if err := c.RemoveJobs(hostID, jobIDs); err != nil {
+		httphelper.NewReponseHelper(w).Error(err)
+		return
+	}
+	w.WriteHeader(200)
+}
+
+func streamHostEvents(c *Cluster, w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	ch := make(chan host.HostEvent)
+	done := make(chan bool)
+	if err := c.StreamHostEvents(ch, done); err != nil {
+		httphelper.NewReponseHelper(w).Error(err)
+		return
+	}
+	go func() {
+		<-w.(http.CloseNotifier).CloseNotify()
+		close(done)
+	}()
+	wr := sse.NewSSEWriter(w)
+	enc := json.NewEncoder(wr)
+	w.Header().Set("Content-Type", "text/event-stream; charset=utf-8")
+	w.WriteHeader(200)
+	wr.Flush()
+	for data := range ch {
+		enc.Encode(data)
+		wr.Flush()
+	}
+}
+
+type ClusterHandle func(*Cluster, http.ResponseWriter, *http.Request, httprouter.Params)
+
+// Helper function for wrapping a ClusterHandle into a httprouter.Handles
+func clusterMiddleware(cluster *Cluster, handle ClusterHandle) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+		handle(cluster, w, r, ps)
+	}
+}
+
+func (cluster *Cluster) ServeHTTP(r *httprouter.Router, sh *shutdown.Handler) error {
+	r.GET("/cluster/hosts", clusterMiddleware(cluster, listHosts))
+	r.PUT("/cluster/hosts/:id", clusterMiddleware(cluster, registerHost))
+	r.POST("/cluster/jobs", clusterMiddleware(cluster, addJobs))
+	r.DELETE("/cluster/hosts/:host_id/jobs/:job_id", clusterMiddleware(cluster, removeJob))
+	r.GET("/cluster/events", clusterMiddleware(cluster, streamHostEvents))
+	return nil
```

Comment:
Could define `clusterMiddleware` within `ServeHTTP` and close over `cluster` to avoid passing it as an arg